File: lib/git-support/add-workflow.rb
require 'json'
require 'net/http'
require 'fileutils'
require 'open3'
require 'securerandom'
require 'colorize'
require 'yaml'

class AddWorkflow
  LOG_FILE = 'logs/add_workflow.log'.freeze

  def initialize(prompt)
    @prompt = prompt
    FileUtils.mkdir_p('logs')
  end

  def run
    config_mode = @prompt.yes?('Do you want to use a config.yml file?')

    config_mode ? load_config : manual_input

    return if !@dry_run && !validate_token

    unless File.exist?(@repo_list_path)
      puts
      puts "Repo list file not found: #{@repo_list_path}".red
      return
    end

    unless File.exist?(@workflow_template_path)
      puts
      puts "Workflow template not found: #{@workflow_template_path}".red
      return
    end

    workflow_filename = File.basename(@workflow_template_path)
    workflow_path = ".github/workflows/#{workflow_filename}"

    repos = File.readlines(@repo_list_path, chomp: true)

    puts
    puts "Found #{repos.count} repositories".green

    repos.each_with_index do |repo, index|
      repo = repo.strip

      next if repo.empty?
      next if repo.start_with?('#')

      puts
      puts "[#{index + 1}/#{repos.count}] Processing repo: #{@org}/#{repo}".light_blue

      log("Processing #{@org}/#{repo}")

      branch_name = build_branch_name

      if @dry_run
        dry_run_output(repo, branch_name, workflow_path)
        next
      end

      process_repository(
        repo: repo,
        workflow_filename: workflow_filename,
        workflow_path: workflow_path,
        branch_name: branch_name
      )
    end

    puts
    puts 'Finished processing repositories'.green
  end

  private

  def manual_input
    @org = ask('GitHub organization')
    @github_token = ask('GitHub token')
    @repo_list_path = ask('Path to repo list file')
    @workflow_template_path = ask('Path to workflow template')
    @branch_suffix = ask('Branch suffix')
    @dry_run = @prompt.yes?('Do you want to run in dry-run mode?')
  end

  def load_config
    config_path = ask('Path to config.yml')

    unless File.exist?(config_path)
      puts
      puts "Config file not found: #{config_path}".red
      exit
    end

    config = YAML.load_file(config_path)

    @org = config['org']
    @github_token = config['github_token']
    @repo_list_path = config['repo_list_path']
    @workflow_template_path = config['workflow_template_path']
    @branch_suffix = config['branch_suffix']
    @dry_run = config['dry_run'] || false

    puts
    puts 'Configuration loaded successfully'.green
  end

  def ask(question)
    loop do
      answer = @prompt.ask("#{question}: ")

      return answer.strip unless answer.nil? || answer.strip.empty?

      print "\e[1A"
      print "\e[2K"

      puts "#{question}: #{'This field cannot be empty'.red}"
    end
  end

  def validate_token
    puts
    puts 'Validating GitHub token...'.cyan

    uri = URI('https://api.github.com/user')

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true

    request = Net::HTTP::Get.new(uri.path)
    request['Authorization'] = "token #{@github_token}"

    response = http.request(request)

    if response.code.to_i == 200
      puts 'GitHub token validated successfully'.green
      true
    else
      puts 'Invalid GitHub token'.red
      false
    end
  end

  def build_branch_name
    "add-workflow-#{@branch_suffix}-#{Time.now.to_i}-#{SecureRandom.hex(4)}"
  end

  def dry_run_output(repo, branch_name, workflow_path)
    puts "[DRY RUN] Would clone #{@org}/#{repo}".yellow
    puts "[DRY RUN] Would create branch #{branch_name}".yellow
    puts "[DRY RUN] Would add #{workflow_path}".yellow
    puts '[DRY RUN] Would commit, push, and open PR'.yellow
  end

  def process_repository(repo:, workflow_filename:, workflow_path:, branch_name:)
    FileUtils.rm_rf(repo)

    begin
      with_retry do
        clone_repository(repo)
      end

      Dir.chdir(repo) do
        default_branch = default_branch_name

        with_retry do
          create_branch(branch_name, default_branch)
        end

        FileUtils.mkdir_p(File.dirname(workflow_path))

        FileUtils.cp(
          File.expand_path(@workflow_template_path, '..'),
          workflow_path
        )

        execute("git add #{workflow_path}")
        execute("git commit -m \"Add workflow #{workflow_filename}\"")
        execute("git push origin #{branch_name}")

        create_pull_request(
          repo: repo,
          branch_name: branch_name,
          default_branch: default_branch,
          workflow_filename: workflow_filename
        )
      end

      puts "Finished #{@org}/#{repo}".green
      log("SUCCESS #{@org}/#{repo}")
    rescue StandardError => e
      puts
      puts "Error processing #{repo}".red
      puts e.message.red

      log("ERROR #{@org}/#{repo} - #{e.message}")
    ensure
      Dir.chdir('..') if Dir.pwd.end_with?(repo)
      FileUtils.rm_rf(repo)
    end
  end

  def execute(command)
    puts
    puts "Executing: #{command}".cyan

    stdout, stderr, status = Open3.capture3(command)

    unless status.success?
      puts stderr.red
      raise "Command failed: #{command}"
    end

    stdout
  end

  def clone_repository(repo)
    execute("git clone https://github.com/#{@org}/#{repo}.git")
  end

  def default_branch_name
    `git remote show origin | awk '/HEAD branch/ {print $NF}'`.strip
  end

  def create_branch(branch_name, default_branch)
    execute("git checkout -b #{branch_name} origin/#{default_branch}")
  end

  def create_pull_request(repo:, branch_name:, default_branch:, workflow_filename:)
    return dry_run_pull_request(repo, branch_name) if @dry_run

    response = github_pull_request_request(
      repo: repo,
      branch_name: branch_name,
      default_branch: default_branch,
      workflow_filename: workflow_filename
    )

    handle_pull_request_response(response, repo)
  end

  def github_pull_request_request(repo:, branch_name:, default_branch:, workflow_filename:)
    uri = URI("https://api.github.com/repos/#{@org}/#{repo}/pulls")

    request = Net::HTTP::Post.new(uri.path)

    request['Authorization'] = "token #{@github_token}"
    request['Accept'] = 'application/vnd.github+json'
    request['Content-Type'] = 'application/json'

    request.body = pull_request_payload(
      branch_name: branch_name,
      default_branch: default_branch,
      workflow_filename: workflow_filename
    ).to_json

    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true

    http.request(request)
  end

  def pull_request_payload(branch_name:, default_branch:, workflow_filename:)
    {
      title: "Add workflow: #{workflow_filename}",
      head: branch_name,
      base: default_branch,
      body: pull_request_body(workflow_filename)
    }
  end

  def pull_request_body(workflow_filename)
    <<~BODY
      Add workflow #{workflow_filename} via automation.

      Generated automatically by script-station.
    BODY
  end

  def handle_pull_request_response(response, repo)
    body = JSON.parse(response.body)

    return log_pull_request_success(body) if body['html_url']

    handle_pull_request_error(body, response, repo)
  end

  def log_pull_request_success(body)
    puts
    puts "→ PR created: #{body['html_url']}".green

    log("PR CREATED #{body['html_url']}")
  end

  def handle_pull_request_error(body, response, repo)
    puts
    if pull_request_already_exists?(body)
      puts '→ Pull request already exists'.yellow

      log("PR ALREADY EXISTS #{@org}/#{repo}")
    else
      puts '→ Failed to create PR'.red
      puts response.body.red

      log("FAILED PR #{@org}/#{repo}")
    end
  end

  def pull_request_already_exists?(body)
    body['errors']&.any? do |error|
      error['message']&.include?('A pull request already exists')
    end
  end

  def dry_run_pull_request(repo, branch_name)
    puts "[DRY RUN] Would create PR for #{repo} from #{branch_name}".yellow
  end

  def with_retry(max_attempts = 3)
    attempts = 0

    begin
      attempts += 1
      yield
    rescue StandardError => e
      if attempts < max_attempts
        puts
        puts "Retrying... Attempt #{attempts}/#{max_attempts}".yellow
        sleep(2)
        retry
      end

      raise e
    end
  end

  def log(message)
    File.open(LOG_FILE, 'a') do |file|
      file.puts("[#{Time.now}] #{message}")
    end
  end
end

File: scripts/add_workflow.py
import base64
import os
import secrets
import time
from datetime import (
    UTC,
    datetime,
)
from pathlib import Path

import questionary
import requests
import yaml

PROJECT_ROOT = (
    Path(__file__)
    .resolve()
    .parent
    .parent
)

LOG_FILE = (
    PROJECT_ROOT
    / "logs"
    / "add_workflow.log"
)


class AddWorkflow:

    def __init__(self):
        self.org = ""
        self.github_token = ""
        self.repo_list_path = None
        self.workflow_template_path = None
        self.branch_suffix = ""
        self.dry_run = False

    def run(self):
        use_config = (
            confirm_action(
                "Do you want to use "
                "a config.yml file?"
            )
        )

        if use_config:
            config_path = (
                ask_required(
                    "Path to config.yml"
                )
            )

            if (
                not config_path
                or not self.load_config(
                    config_path
                )
            ):
                return 1

        elif not self.manual_input():
            return 1

        if (
            not self.dry_run
            and not self.validate_token()
        ):
            return 1

        if not self.validate_files():
            return 1

        repos = (
            self.load_repositories()
        )

        print(
            "\nFound "
            f"{len(repos)} repositories"
        )

        failed = False

        for index, repo in enumerate(
            repos,
            start=1,
        ):
            print(
                "\n"
                f"[{index}/{len(repos)}] "
                "Processing repo: "
                f"{self.org}/{repo}"
            )

            self.log(
                "Processing "
                f"{self.org}/{repo}"
            )

            branch_name = (
                self.build_branch_name()
            )

            workflow_filename = (
                self
                .workflow_template_path
                .name
            )

            workflow_path = (
                ".github/workflows/"
                f"{workflow_filename}"
            )

            if self.dry_run:
                self.dry_run_output(
                    repo,
                    branch_name,
                    workflow_path,
                )

                continue

            try:
                self.process_repository(
                    repo,
                    workflow_filename,
                    workflow_path,
                    branch_name,
                )

            except (
                requests.RequestException,
                RuntimeError,
                ValueError,
            ) as error:
                failed = True

                print(
                    "\nError processing "
                    f"{repo}: {error}"
                )

                self.log(
                    "ERROR "
                    f"{self.org}/{repo} - "
                    f"{error}"
                )

        print(
            "\nFinished processing "
            "repositories"
        )

        return (
            1
            if failed
            else 0
        )

    def manual_input(self):
        self.org = ask_required(
            "GitHub organization"
        )

        repo_list_path = ask_required(
            "Path to repo list file"
        )

        workflow_template_path = (
            ask_required(
                "Path to workflow template"
            )
        )

        self.branch_suffix = (
            ask_required(
                "Branch suffix"
            )
        )

        if not all(
            (
                self.org,
                repo_list_path,
                workflow_template_path,
                self.branch_suffix,
            )
        ):
            return False

        self.repo_list_path = (
            resolve_path(
                repo_list_path
            )
        )

        self.workflow_template_path = (
            resolve_path(
                workflow_template_path
            )
        )

        self.dry_run = (
            confirm_action(
                "Do you want to run "
                "in dry-run mode?"
            )
        )

        if not self.dry_run:
            self.github_token = (
                os.getenv(
                    "GITHUB_TOKEN",
                    "",
                )
                .strip()
            )

            if not self.github_token:
                self.github_token = (
                    ask_required(
                        "GitHub token",
                        secret=True,
                    )
                    or ""
                )

        return True

    def load_config(
        self,
        config_path,
    ):
        path = resolve_path(
            config_path
        )

        if not path.exists():
            print(
                "\nConfig file not found: "
                f"{path}"
            )

            return False

        with path.open(
            "r",
            encoding="utf-8",
        ) as file:
            config = (
                yaml.safe_load(
                    file
                )
                or {}
            )

        self.org = str(
            config.get(
                "org",
                "",
            )
        ).strip()

        repo_list_path = str(
            config.get(
                "repo_list_path",
                "",
            )
        ).strip()

        workflow_template_path = str(
            config.get(
                "workflow_template_path",
                "",
            )
        ).strip()

        self.branch_suffix = str(
            config.get(
                "branch_suffix",
                "",
            )
        ).strip()

        self.dry_run = bool(
            config.get(
                "dry_run",
                False,
            )
        )

        token_env = str(
            config.get(
                "github_token_env",
                "GITHUB_TOKEN",
            )
        ).strip()

        self.github_token = str(
            config.get(
                "github_token",
                "",
            )
        ).strip()

        if not self.github_token:
            self.github_token = (
                os.getenv(
                    token_env,
                    "",
                )
                .strip()
            )

        if repo_list_path:
            self.repo_list_path = (
                resolve_path(
                    repo_list_path
                )
            )

        if workflow_template_path:
            self.workflow_template_path = (
                resolve_path(
                    workflow_template_path
                )
            )

        missing = []

        if not self.org:
            missing.append(
                "org"
            )

        if not repo_list_path:
            missing.append(
                "repo_list_path"
            )

        if not workflow_template_path:
            missing.append(
                "workflow_template_path"
            )

        if not self.branch_suffix:
            missing.append(
                "branch_suffix"
            )

        if (
            not self.dry_run
            and not self.github_token
        ):
            missing.append(
                token_env
            )

        if missing:
            print(
                "\nMissing configuration:"
            )

            for name in missing:
                print(
                    f"  - {name}"
                )

            return False

        print(
            "\nConfiguration loaded "
            "successfully"
        )

        return True

    def validate_token(self):
        print(
            "\nValidating GitHub token..."
        )

        response = requests.get(
            "https://api.github.com/user",
            headers=self.github_headers(),
            timeout=30,
        )

        if response.status_code == 200:
            print(
                "GitHub token validated "
                "successfully"
            )

            return True

        print(
            "Invalid GitHub token"
        )

        return False

    def validate_files(self):
        if (
            not self.repo_list_path
            or not self
            .repo_list_path
            .exists()
        ):
            print(
                "\nRepo list file not found: "
                f"{self.repo_list_path}"
            )

            return False

        if (
            not self.workflow_template_path
            or not self
            .workflow_template_path
            .exists()
        ):
            print(
                "\nWorkflow template "
                "not found: "
                f"{self.workflow_template_path}"
            )

            return False

        return True

    def load_repositories(self):
        repos = []

        lines = (
            self.repo_list_path
            .read_text(
                encoding="utf-8"
            )
            .splitlines()
        )

        for raw_line in lines:
            line = raw_line.strip()

            if (
                line
                and not line.startswith(
                    "#"
                )
            ):
                repos.append(
                    line
                )

        return repos

    def build_branch_name(self):
        return (
            "add-workflow-"
            f"{self.branch_suffix}-"
            f"{int(time.time())}-"
            f"{secrets.token_hex(4)}"
        )

    def dry_run_output(
        self,
        repo,
        branch_name,
        workflow_path,
    ):
        print(
            "[DRY RUN] Would update "
            f"{self.org}/{repo}"
        )

        print(
            "[DRY RUN] Would create "
            f"branch {branch_name}"
        )

        print(
            "[DRY RUN] Would add "
            f"{workflow_path}"
        )

        print(
            "[DRY RUN] Would commit "
            "and open a pull request"
        )

    def process_repository(
        self,
        repo,
        workflow_filename,
        workflow_path,
        branch_name,
    ):
        default_branch = (
            self.get_default_branch(
                repo
            )
        )

        base_sha = (
            self.get_branch_sha(
                repo,
                default_branch,
            )
        )

        self.create_branch(
            repo,
            branch_name,
            base_sha,
        )

        self.upload_workflow(
            repo,
            branch_name,
            workflow_path,
            workflow_filename,
        )

        self.create_pull_request(
            repo,
            branch_name,
            default_branch,
            workflow_filename,
        )

        print(
            "Finished "
            f"{self.org}/{repo}"
        )

        self.log(
            "SUCCESS "
            f"{self.org}/{repo}"
        )

    def get_default_branch(
        self,
        repo,
    ):
        response = self.request(
            "GET",
            repo,
        )

        default_branch = (
            response
            .json()
            .get(
                "default_branch"
            )
        )

        if not default_branch:
            raise RuntimeError(
                "Default branch not found"
            )

        return default_branch

    def get_branch_sha(
        self,
        repo,
        branch,
    ):
        response = self.request(
            "GET",
            repo,
            (
                "git/ref/heads/"
                f"{branch}"
            ),
        )

        sha = (
            response
            .json()
            .get(
                "object",
                {},
            )
            .get(
                "sha"
            )
        )

        if not sha:
            raise RuntimeError(
                "Default branch SHA "
                "not found"
            )

        return sha

    def create_branch(
        self,
        repo,
        branch_name,
        base_sha,
    ):
        self.request(
            "POST",
            repo,
            "git/refs",
            data={
                "ref":
                    (
                        "refs/heads/"
                        f"{branch_name}"
                    ),
                "sha":
                    base_sha,
            },
            expected=(
                201,
            ),
        )

    def upload_workflow(
        self,
        repo,
        branch_name,
        workflow_path,
        workflow_filename,
    ):
        content = (
            base64
            .b64encode(
                self
                .workflow_template_path
                .read_bytes()
            )
            .decode(
                "utf-8"
            )
        )

        self.request(
            "PUT",
            repo,
            (
                "contents/"
                f"{workflow_path}"
            ),
            data={
                "message":
                    (
                        "Add workflow "
                        f"{workflow_filename}"
                    ),
                "content":
                    content,
                "branch":
                    branch_name,
            },
            expected=(
                201,
            ),
        )

    def create_pull_request(
        self,
        repo,
        branch_name,
        default_branch,
        workflow_filename,
    ):
        response = requests.post(
            (
                "https://api.github.com/"
                "repos/"
                f"{self.org}/{repo}/pulls"
            ),
            headers=self.github_headers(),
            json={
                "title":
                    (
                        "Add workflow: "
                        f"{workflow_filename}"
                    ),
                "head":
                    branch_name,
                "base":
                    default_branch,
                "body":
                    (
                        "Add workflow "
                        f"{workflow_filename} "
                        "via automation.\n\n"
                        "Generated automatically "
                        "by script-station."
                    ),
            },
            timeout=30,
        )

        if response.status_code == 201:
            url = (
                response
                .json()
                .get(
                    "html_url"
                )
            )

            print(
                f"→ PR created: {url}"
            )

            self.log(
                f"PR CREATED {url}"
            )

            return

        body = safe_json(
            response
        )

        if pull_request_exists(
            body
        ):
            print(
                "→ Pull request "
                "already exists"
            )

            self.log(
                "PR ALREADY EXISTS "
                f"{self.org}/{repo}"
            )

            return

        raise RuntimeError(
            "Failed to create "
            "pull request: HTTP "
            f"{response.status_code}"
        )

    def request(
        self,
        method,
        repo,
        path="",
        data=None,
        expected=(200,),
    ):
        suffix = (
            f"/{path}"
            if path
            else ""
        )

        response = requests.request(
            method,
            (
                "https://api.github.com/"
                "repos/"
                f"{self.org}/{repo}"
                f"{suffix}"
            ),
            headers=self.github_headers(),
            json=data,
            timeout=30,
        )

        if (
            response.status_code
            not in expected
        ):
            raise RuntimeError(
                "GitHub API "
                f"{method} failed: HTTP "
                f"{response.status_code}"
            )

        return response

    def github_headers(self):
        return {
            "Authorization":
                (
                    "Bearer "
                    f"{self.github_token}"
                ),
            "Accept":
                (
                    "application/vnd."
                    "github+json"
                ),
            "X-GitHub-Api-Version":
                "2022-11-28",
        }

    def log(
        self,
        message,
    ):
        LOG_FILE.parent.mkdir(
            parents=True,
            exist_ok=True,
        )

        with LOG_FILE.open(
            "a",
            encoding="utf-8",
        ) as file:
            file.write(
                "["
                f"{datetime.now(UTC).isoformat()}"
                "] "
                f"{message}\n"
            )


def resolve_path(
    value,
):
    path = (
        Path(
            str(value)
        )
        .expanduser()
    )

    if path.is_absolute():
        return path

    return (
        PROJECT_ROOT
        / path
    ).resolve()


def ask_required(
    message,
    secret=False,
):
    while True:
        prompt = (
            questionary.password
            if secret
            else questionary.text
        )

        answer = (
            prompt(
                f"{message}:"
            )
            .ask()
        )

        if answer is None:
            return None

        answer = answer.strip()

        if answer:
            return answer

        print(
            f"{message}: "
            "This field cannot be empty"
        )


def confirm_action(
    message,
):
    return bool(
        questionary
        .confirm(
            message,
            default=False,
        )
        .ask()
    )


def safe_json(
    response,
):
    try:
        return response.json()

    except ValueError:
        return {}


def pull_request_exists(
    body,
):
    return any(
        (
            "A pull request "
            "already exists"
        )
        in error.get(
            "message",
            "",
        )
        for error in body.get(
            "errors",
            [],
        )
    )


def main():
    try:
        return (
            AddWorkflow()
            .run()
        )

    except KeyboardInterrupt:
        print(
            "\nExiting Add Workflow."
        )

        return 130


if __name__ == "__main__":  # pragma: no cover
    raise SystemExit(
        main()
    )
